hamilton/io/fingerprinting.py

```diff
@@ -0,0 +1,117 @@
+import base64
+import functools
+import hashlib
+import logging
+from collections.abc import Mapping, Sequence, Set
+
+from hamilton.experimental import h_databackends
+
+logger = logging.getLogger(__name__)
+
+
+MAX_DEPTH = 4
+
+
+def _compact_hash(digest: bytes) -> str:
+    """Compact the hash to a string that's safe to pass around.
+
+    NOTE this is particularly relevant for th Hamilton UI and
+    passing hashes/fingerprints through web services.
+    """
+    return base64.urlsafe_b64encode(digest).decode()
+
+
+@functools.singledispatch
+def hash_value(obj, *args, depth=0, **kwargs) -> str:
+    """Fingerprinting strategy that computes a hash of the
+    full Python object.
+
+    The default case hashes the `__dict__` attribute of the
+    object (recursive).
+    """
+    if hasattr(obj, "__dict__") and depth < MAX_DEPTH:
+        depth += 1
+        return hash_value(obj.__dict__, depth=depth)
+
+    hash_object = hashlib.md5("<unhashable>".encode())
+    return _compact_hash(hash_object.digest())
+
+
+@hash_value.register(str)
+@hash_value.register(int)
+@hash_value.register(float)
+@hash_value.register(bool)
+@hash_value.register(bytes)
+def hash_primitive(obj, *args, **kwargs) -> str:
+    """Convert the primitive to a string and hash it"""
+    hash_object = hashlib.md5(str(obj).encode())
```

Comment:
This conversion is need because `hash_primitive` is registered for `str`, `int`, `float`,  `bool`, which don't have an `.encode()` method.

I will create a separate method for `bytes`